agent/consul/state/catalog_events.go

```diff
@@ -0,0 +1,465 @@
+				// Service was renamed, the code below will ensure the new registrations
+				// go out to subscribers to the new service name topic key, but we need
+				// to fix up subscribers that were watching the old name by sending
+				// deregistrations.
+				e := newServiceHealthEventDeregister(changes.Index, before)
+				events = append(events, e)
+			}
+
+			if before.ServiceKind == structs.ServiceKindConnectProxy &&
+				before.ServiceProxy.DestinationServiceName != after.ServiceProxy.DestinationServiceName {
+				// Connect proxy changed the service it is representing, need to issue a
+				// dereg for the old service on the Connect topic. We don't actually need
+				// to deregister this sidecar service though as it still exists and
+				// didn't change its name (or if it did that was caught just above). But
+				// our mechanism for connect events is to convert them so we generate
+				// the regular one, convert it to Connect topic and then discard the
+				// original.
+				e := newServiceHealthEventDeregister(changes.Index, before)
+				events = append(events, serviceHealthToConnectEvents(e)...)
+			}
+		}
+
+		if _, ok := nodeChanges[tuple.Node]; ok {
+			// We already rebuilt events for everything on this node, no need to send
+			// a duplicate.
+			continue
+		}
+		// Build service event and append it
+		e, err := newServiceHealthEventForService(tx, changes.Index, tuple)
+		if err != nil {
+			return nil, err
+		}
+		events = append(events, e)
+	}
+
+	// Duplicate any events that affected connect-enabled instances (proxies or
+	// native apps) to the relevant Connect topic.
+	events = append(events, serviceHealthToConnectEvents(events...)...)
+
+	return events, nil
+}
+
+type changeType uint8
+
+const (
+	// changeIndirect indicates some other object changed which has implications
+	// for the target object.
+	changeIndirect changeType = iota
+	changeDelete
+	changeCreate
+	changeUpdate
+)
+
+func changeTypeFromChange(change memdb.Change) changeType {
+	switch {
+	case change.Deleted():
+		return changeDelete
+	case change.Created():
+		return changeCreate
+	default:
+		return changeUpdate
+	}
+}
+
+// serviceHealthToConnectEvents converts already formatted service health
+// registration events into the ones needed to publish to the Connect topic.
+// This essentially means filtering out any instances that are not Connect
+// enabled and so of no interest to those subscribers but also involves
+// switching connection details to be the proxy instead of the actual instance
+// in case of a sidecar.
+func serviceHealthToConnectEvents(events ...stream.Event) []stream.Event {
+	var serviceHealthConnectEvents []stream.Event
+	for _, event := range events {
+		if event.Topic != TopicServiceHealth {
+			// Skip non-health or any events already emitted to Connect topic
+			continue
+		}
+		node := getPayloadCheckServiceNode(event.Payload)
+		// TODO: do we need to handle gateways here as well?
```

Comment:
We do need terminating gateways here but not mesh or ingress.

The reason is described in the function comment but roughly: Connect discovery should return all connect-enabled endpoints for a service.

 - Termiating gateways _do_ expose "connect enabled" listeners to proxy to external services so should be included
 - Mesh gateways are never exposing connect certs directly just plumbing between connect enabled endpoints.
 - ingress gateways only expose non-connect listeners to external clients by definition so don't have any thing to return to connect-enabled workloads.

Reply:
Makes sense! I've added Terminating Gateways here, but there is a second special case just below that sets the key to the DestinationServiceName. I've added another TODO there to confirm we don't need to do anything special for terminating gateways.